### ISSUE TITLE
How to formulate training loop via torch device API

### ISSUE THREAD

**Issue by nsmithtt** [tenstorrent/tt-mlir#82]
Running into dynamo errors trying to push through vanilla torch 2.0 training path.  Training construction reference: https://pytorch.org/tutorials/intermediate/torch_compile_tutorial.html#demonstrating-speedups (see code snippet under "Now, let’s consider comparing training.")


```python
class MNIST(nn.Module):
    def __init__(self, input_size=784, output_size=10, hidden_size=256):
        super(MNIST, self).__init__()
        self.l1 = nn.Linear(input_size, hidden_size)
        self.relu = nn.ReLU()
        self.l2 = nn.Linear(hidden_size, output_size)
        
    def forward(self, x):
        x = self.l1(x)
        x = self.relu(x)
        x = self.l2(x)
        return nn.functional.softmax(x)


def test_mnist_training():               
    mnist = MNIST().to("tt")                                              
    loss_fn = torch.nn.CrossEntropyLoss().to("tt")                                                 
    opt = torch.optim.Adam(mnist.parameters())                                                     
                                                                                                   
    def train(mod, data):                                  
        opt.zero_grad(True) 
        pred = mod(data[0])
        loss = loss_fn(pred, data[1])                                                                                                                                                                                 
        loss.backward()                                                                                                                                                                                               
        opt.step()                                                                                                                                                                                                    

    batch = 1
    input_img = torch.randn(batch, 28*28)
    train_opt = torch.compile(train, backend=compile_torch, dynamic=False)
    target = nn.functional.one_hot(torch.randint(0, 10, (batch,)), num_classes=10).float().to("tt")
    result = train_opt(mnist, [input_img.to("tt"), target])
    print(result)
```

**Comment by nvukobratTT** [comment]
Summary of sub-issues related to training APIs.

Op level support:

- [Removal of forward graph softmax decomposition on PyBuda](https://github.com/tenstorrent/tt-mlir/issues/77)
- [Initial set of ops required to run e2e (TVM -> PyBuda Compile -> MLIR -> Flattbuffer -> PyBuda Runtime)](https://github.com/tenstorrent/tt-mlir/issues/145)

Inference support:

- [Train MNIST based Linear model on CPU](https://github.com/tenstorrent/tt-mlir/issues/150) (model arch validation)
- [MNIST based Liner model - inference e2e](https://github.com/tenstorrent/tt-mlir/issues/149)

Training support:

- [Training infrastructure (record/track loss, gradients, accuracy, etc)](https://github.com/tenstorrent/tt-mlir/issues/151)
- [Running MNIST forward, loss and backward graph (without optimizer)](https://github.com/tenstorrent/tt-mlir/issues/186)
- [Optimizer support - updating model parameters](https://github.com/tenstorrent/tt-mlir/issues/183)
- [Ability to move inputs/parameters between memory locations (e.g. host, device)](https://github.com/tenstorrent/tt-mlir/issues/154)
- [Support of loss.backward() e2e - running model backward graph](https://github.com/tenstorrent/tt-mlir/issues/182)
- [Training MNIST based Linear model e2e](https://github.com/tenstorrent/tt-mlir/issues/156) - encapsulate core details of each training component

@nsmithtt @staylorTT having issues aligned above, do you still see value of having this one opened, or we're ready to close it?

**Comment by nsmithtt** [comment]
I think we can close this one.  Thank you for driving all of this!

Btw for itemized issue lists like you have above, GitHub has special support for list elements of linked issues, so if you write:
```markdown
- #77
```

You get:
- #77

And the status will automatically be checked when the issue is closed:
- #87

**Comment by nvukobratTT** [comment]
As we created new training milestone based on effort of this issue, I'm going to close it.
